### ISSUE TITLE
Support millisecond time units in interval parsing

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#6420]
This would allow us to migrate off of the `parse_duration` crate (#6244).

The code in question is here: https://github.com/MaterializeInc/materialize/blob/0b0e493a15dc06aac079a6b53feada0ddb085a92/src/repr/src/strconv.rs#L388-L418

/cc @justinj @sploiselle

**Comment by jkosh44** [comment]
Milliseconds and microseconds are handled mostly the same, so I'm going to include microseconds in this.

Postgres documentation for interval parsing can be found [here](https://www.postgresql.org/docs/14/datatype-datetime.html#DATATYPE-INTERVAL-INPUT).

The postgres function for parsing an interval can be found here: 
https://github.com/postgres/postgres/blob/5ef1eefd76f404ddc59b885d50340e602b70f05f/src/backend/utils/adt/timestamp.c#L875-L952
Below are some of the helper functions that are called:
ParseDateTime: https://github.com/postgres/postgres/blob/5ef1eefd76f404ddc59b885d50340e602b70f05f/src/backend/utils/adt/datetime.c#L550-L775
DecodeInterval: https://github.com/postgres/postgres/blob/5ef1eefd76f404ddc59b885d50340e602b70f05f/src/backend/utils/adt/datetime.c#L3084-L3453
DecodeISO8601Interval: https://github.com/postgres/postgres/blob/5ef1eefd76f404ddc59b885d50340e602b70f05f/src/backend/utils/adt/datetime.c#L3500-L3713
AdjustIntervalForTypmod: https://github.com/postgres/postgres/blob/5ef1eefd76f404ddc59b885d50340e602b70f05f/src/backend/utils/adt/timestamp.c#L1325-L1491

Postgres supports 3 different syntaxes for Intervals. Below I will give some details on the three and explain how our handling of millisecond/microsecond diverge.

# SQL Standard Format
This has the format `<year>-<month> <day> <hour>:<minute>:<second>`. The way to include milliseconds/microseconds is to add a decimal component to the second field. Materialize handles this properly.
```
postgres=# SELECT interval '1-2 3 4:05:06.007008';
              interval               
-------------------------------------
 1 year 2 mons 3 days 04:05:06.007008
(1 row)
```
```
materialize=> SELECT interval '1-2 3 4:05:06.007008';
               interval                
---------------------------------------
 1 year 2 months 3 days 04:05:06.007008
(1 row)
```

# Postgres Verbose Format
This has the format: `P quantity unit [ quantity unit ...] [ T [ quantity unit ...]]`. There are two ways to include milliseconds/microseconds.
1. Include it in the decimal component of the second field. Materialize handles this properly.
```
postgres=# SELECT interval '1 year 2 months 3 days 4 hours 5 minutes 6.007008 seconds';
               interval               
--------------------------------------
 1 year 2 mons 3 days 04:05:06.007008
(1 row)
```
```
materialize=> SELECT interval '1 year 2 months 3 days 4 hours 5 minutes 6.007008 seconds';
                interval                
----------------------------------------
 1 year 2 months 3 days 04:05:06.007008
(1 row)
```
2. Include millisecond and microsecond units. Materialize DOES NOT handle this properly.
```
postgres=# SELECT interval '1 year 2 months 3 days 4 hours 5 minutes 6 seconds 7 milliseconds 8 microseconds';
               interval               
--------------------------------------
 1 year 2 mons 3 days 04:05:06.007008
(1 row)
```
```
materialize=> SELECT interval '1 year 2 months 3 days 4 hours 5 minutes 6 seconds 7 milliseconds 8 microseconds';
ERROR:  invalid input syntax for type interval: Cannot determine format of all parts. Add explicit time components, e.g. INTERVAL '1 day' or INTERVAL '1' DAY: "1 year 2 months 3 days 4 hours 5 minutes 6 seconds 7 milliseconds 8 microseconds"
```
**NOTE**: Materialize also does not support the following units that postgres does support:
- `week`
- `decade`
- `century`
- `millennium`
If it's trivial to add them then I'll add them to the fix of this issue. Otherwise I'll open a separate issue.

# ISO 8601 Format
This format contains 2 different variants.
## With Designators
This has the format: `P quantity unit [ quantity unit ...] [ T [ quantity unit ...]]`. The way to include milliseconds/microseconds is to add a decimal component to the second unit. There is no millisecond/microsecond unit. Materialize DOES NOT support this format for intervals.
```
postgres=# SELECT interval 'P1Y2M3DT4H5M6.007008S';
               interval               
--------------------------------------
 1 year 2 mons 3 days 04:05:06.007008
(1 row)
```
```
materialize=> SELECT interval 'P1Y2M3DT4H5M6.007008S';
ERROR:  invalid input syntax for type interval: unknown units P: "P1Y2M3DT4H5M6.007008S"
```
## Alternative Format
This has the format: `P [ years-months-days ] [ T hours:minutes:seconds ]`. The way to include milliseconds/microseconds is to add a decimal component to the second field. Materialize DOES NOT support this format for intervals.
```
postgres=# SELECT interval 'P01-02-03T04:05:06.007008';
               interval               
--------------------------------------
 1 year 2 mons 3 days 04:05:06.007008
(1 row)
```
```
materialize=> SELECT interval 'P01-02-03T04:05:06.007008';
ERROR:  invalid input syntax for type interval: unknown units P: "P01-02-03T04:05:06.007008"
```

# Type Modifiers
Intervals can also include an optional Type Modifier, which is placed after the interval type. The Type Modifier acts very similarly to the `date_trunc` function. millisecond and microsecond are not valid type modifiers. Materialize behaves the same way as postgres with type modifiers.

```
postgres=# SELECT interval '01-02 03 04:05:06.007008' hour;
           interval            
-------------------------------
 1 year 2 mons 3 days 04:00:00
(1 row)
```
```
materialize=> SELECT interval '01-02 03 04:05:06.007008' hour;
            interval             
---------------------------------
 1 year 2 months 3 days 04:00:00
(1 row)
```
Note for the following two examples `millisecond` is treated as a column alias since it's not a valid Type Modifier.
```
postgres=# SELECT interval '01-02 03 04:05:06.007008' millisecond;
             millisecond              
--------------------------------------
 1 year 2 mons 3 days 04:05:06.007008
(1 row)
```
```
materialize=> SELECT interval '01-02 03 04:05:06.007008' millisecond;
              millisecond               
----------------------------------------
 1 year 2 months 3 days 04:05:06.007008
(1 row)
```

# TLDR
- To fix this issue `millisecond` and `microsecond` need to be added as units to the Postgres Verbose Format.
- If it's trivial to implement, the following units also need to be added to the Postgres Verbose Format: `week`, `decade`, `century`, `millennium`. Otherwise a separate issue will be opened.
- A separate issue will be opened to implement the ISO 8601 formats.

**Comment by jkosh44** [comment]
One relevant corner case is that PostgreSQL does not allow you to specify a decimal second and millisecond/microsecond unit in the Postgres Verbose Format.

```
postgres=# SELECT interval '6.007 seconds 9 milliseconds';
ERROR:  invalid input syntax for type interval: "6.007 seconds 9 milliseconds"
LINE 1: SELECT interval '6.007 seconds 9 milliseconds';
                        ^
postgres=# SELECT interval '6.007 seconds 9 microseconds';
ERROR:  invalid input syntax for type interval: "6.007 seconds 9 microseconds"
LINE 1: SELECT interval '6.007 seconds 9 microseconds';
```

**Comment by sploiselle** [comment]
@jkosh44 That makes sense; you can only specify each "component" of the interval once, so it likely blocks all subsecond classes if you set any of them in the second. Would be interested to see the behavior if they're explicitly set to e.g. 0.

**Comment by jkosh44** [comment]
> @jkosh44 That makes sense; you can only specify each "component" of the interval once, so it likely blocks all subsecond classes if you set any of them in the second. Would be interested to see the behavior if they're explicitly set to e.g. 0.

```
postgres=# SELECT interval '3.00 seconds 7ms 6us';
    interval     
-----------------
 00:00:03.007006
(1 row)
```

Looks like that is allowed.
